src/middlewared/middlewared/plugins/virt/instance.py

```diff
@@ -0,0 +1,265 @@
+import aiohttp
+
+from middlewared.service import (
+    CRUDService, ValidationErrors, filterable, job, private
+)
+from middlewared.utils import filter_list
+
+from middlewared.api import api_method
+from middlewared.api.current import (
+    VirtInstanceEntry,
+    VirtInstanceCreateArgs, VirtInstanceCreateResult,
+    VirtInstanceUpdateArgs, VirtInstanceUpdateResult,
+    VirtInstanceDeleteArgs, VirtInstanceDeleteResult,
+    VirtInstanceStartArgs, VirtInstanceStartResult,
+    VirtInstanceStopArgs, VirtInstanceStopResult,
+    VirtInstanceRestartArgs, VirtInstanceRestartResult,
+    VirtInstanceImageChoicesArgs, VirtInstanceImageChoicesResult,
+)
+from .utils import incus_call, incus_call_and_wait
+
+
+LC_IMAGES_SERVER = 'https://images.linuxcontainers.org'
+LC_IMAGES_JSON = f'{LC_IMAGES_SERVER}/streams/v1/images.json'
+
+
+class VirtInstanceService(CRUDService):
+
+    class Config:
+        namespace = 'virt.instance'
+        cli_namespace = 'virt.instance'
+        entry = VirtInstanceEntry
+        role_prefix = 'VIRT_INSTANCE'
+
+    @filterable
+    async def query(self, filters, options):
+        """
+        Query all instances with `query-filters` and `query-options`.
+        """
+        results = (await incus_call('1.0/instances?filter=&recursion=2', 'get'))['metadata']
+        entries = []
+        for i in results:
+            # If entry has no config or state its probably in an unknown state, skip it
+            if not i.get('config') or not i.get('state'):
+                continue
+            entry = {
+                'id': i['name'],
+                'name': i['name'],
+                'type': 'CONTAINER' if i['type'] == 'container' else 'VM',
+                'status': i['state']['status'].upper(),
+                'cpu': i['config'].get('limits.cpu'),
+                'autostart': i['config'].get('boot.autostart') or False,
+                'environment': {},
+                'raw': i,
```

Comment:
Also i think there is no need to send `raw` value here and instead if we really want it - have an extra argument for it to be there because it is noise otherwise?